### PR TITLE
added matrix factorization test

### PR DESCRIPTION
**Pull request by cem3394**
Added a simple test that factors a 2x2 matrix of rank 1 into the outer product of two vectors.

### CODE REVIEW

File: tensorflow-ops/tests/MatrixTest.hs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import qualified Data.Vector as V
+
+import Test.Framework (Test)
+import Test.Framework.Providers.HUnit (testCase)
+import TensorFlow.Test (assertAllClose)
+import Google.Test (googleTest)
+
+
+randomParam (TF.Shape shape) =
+  (`TF.mul` (TF.scalar 1.0)) <$> TF.truncatedNormal (TF.vector shape)
```
Comment (fkm3) [inline tensorflow-ops/tests/MatrixTest.hs]: Can shorten this to `TF.truncatedNormal (TF.vector shape)`, right?

File: tensorflow-ops/tests/MatrixTest.hs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+reduceMean :: TF.Tensor TF.Build Float -> TF.Tensor TF.Build Float
+reduceMean xs = TF.mean xs (TF.scalar (0 :: Int32))
```
Comment (fkm3) [inline tensorflow-ops/tests/MatrixTest.hs]: Here is a better implementation so you don't have to do `reduceMean . reduceMean`:

```haskell
reduceMean :: TF.Tensor TF.Value Float -> TF.Tensor TF.Value Float
reduceMean xs = TF.mean xs (TF.range 0 (TF.rank xs) 1)
```

Maybe add it to `TensorFlow.Ops`

File: tensorflow-ops/tests/MatrixTest.hs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+fitMatrix :: Test
+fitMatrix = testCase "fitMatrix" $ TF.runSession $ do
+  u <- TF.initializedVariable =<< randomParam [2, 1]
+  v <- TF.initializedVariable =<< randomParam [1, 2]
+  let ones = [1, 1, 1, 1] :: [Float]
+      matx = TF.constant [2, 2] ones
+      diff = matx `TF.sub` (u `TF.matMul` v)
+      loss = reduceMean . reduceMean $ square diff
+  trainStep <- gradientDescent 0.01 loss [u, v]
+  replicateM_ 1000 (TF.run trainStep)
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/MatrixTest.hs]: Can this be done in fewer than 1000 steps? It seems pretty sad that such a simple optimization is taking this long.

File: tensorflow-ops/tests/MatrixTest.hs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import Google.Test (googleTest)
+
+
+randomParam (TF.Shape shape) =
+  (`TF.mul` (TF.scalar 1.0)) <$> TF.truncatedNormal (TF.vector shape)
+
+reduceMean :: TF.Tensor TF.Build Float -> TF.Tensor TF.Build Float
+reduceMean xs = TF.mean xs (TF.scalar (0 :: Int32))
+
+fitMatrix :: Test
```
Comment (fkm3) [inline tensorflow-ops/tests/MatrixTest.hs]:
```haskell
-- Simple test that factors a 2x2 matrix of rank 1 into the outer product of two vectors.
fitMatrix :: Test
```

**Comment by fkm3** [comment]
Jenkins, test this please.

File: tensorflow-ops/tests/MatrixTest.hs (hunk 2 above)
Comment (fkm3) [inline tensorflow-ops/tests/MatrixTest.hs]: Actually, it should be more like

```haskell
reduceMean :: TF.Tensor v Float -> TF.Tensor TF.Build Float
reduceMean xs = TF.mean xs (TF.range 0 (TF.rank xs) 1)
```

(wasn't sync'd to HEAD)

File: tensorflow-ops/tests/MatrixTest.hs (hunk 3 above)
Comment (cem3394) [inline tensorflow-ops/tests/MatrixTest.hs]: It's overkill. I added several orders of magnitude more precision than needed to ensure the solution would be within the tolerance level every time.

File: tensorflow-ops/tests/MatrixTest.hs (hunk 1 above)
Comment (cem3394) [inline tensorflow-ops/tests/MatrixTest.hs]: Yes you can, thanks.

**Comment by cem3394** [comment]
Made those 2 changes, 300 iterations is approximately the minimum number required to consistently beat the 0.001 tolerance in `assertAllClose`.

**Comment by blackgnezdo** [comment]
Jenkins, test this please.

**Comment by fkm3** [comment]
Jenkins, test this please.

**Comment by blackgnezdo** [comment]
@cem3394 I imagine it's fairly frustrating to wait for Jenkins to build your code. You can confirm it locally by running stack the same way we do in CI: [stack build --pedantic --test](https://github.com/tensorflow/haskell/blob/master/ci_build/outer_launch_tests.sh)

I also would appreciate if you re-wrote git history with the same email address that you used for signing CLA, squashing and submitting a new PR is a fine alternative.

**Comment by fkm3** [comment]
Jenkins, test this please.
